File: webpack.config.d/dev-credentials.js
// Dev auto-login credentials (see DevAutoLogin.kt) are read from the environment of the
// process that runs webpack and inlined as globals, development bundles only. Production
// bundles always get empty strings, so a deployed bundle can never carry them.
const webpack = require('webpack');
const dev = config.mode !== 'production';
config.plugins.push(new webpack.DefinePlugin({
    __VOID_DEV_USERNAME__: JSON.stringify(dev ? (process.env.VOID_DEV_USERNAME || '') : ''),
    __VOID_DEV_PASSWORD__: JSON.stringify(dev ? (process.env.VOID_DEV_PASSWORD || '') : ''),
}));

File: src/commonMain/kotlin/DevCredentials.kt
/**
 * Developer credential lookup for [DevAutoLogin]. Values live in the environment, never in
 * source: `VOID_DEV_USERNAME` / `VOID_DEV_PASSWORD`. JVM reads the process environment at
 * startup; the browser bundle gets them inlined at webpack time from the environment of the
 * process that ran the build (see webpack.config.d/dev-credentials.js). Blank means unset.
 */
expect fun devCredential(name: String): String?
